src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/ExtractToCodeBehindCodeActionResolver.cs

```diff
@@ -0,0 +1,217 @@
+                            Range = StartOfDocumentRange,
+                        }
+                    },
+                })
+            };
+            
+            return new WorkspaceEdit
+            {
+                DocumentChanges = documentChanges,
+            };
+        }
+
+        /// <summary>
+        /// Generate a file path with adjacent to our input path that has the
+        /// correct codebehind extension, using numbers to differentiate from
+        /// any collisions.
+        /// </summary>
+        /// <param name="path">The origin file path.</param>
+        /// <returns>A non-existent file path with the same base name and a codebehind extension.</returns>
+        private string GenerateCodeBehindPath(string path)
+        {
+            var n = 0;
+            string codeBehindPath;
+            do
+            {
+                var identifier = n > 0 ? n.ToString() : string.Empty;  // Make it look nice
+                codeBehindPath = Path.Combine(
+                    Path.GetDirectoryName(path),
+                    $"{Path.GetFileNameWithoutExtension(path)}{identifier}{Path.GetExtension(path)}.cs");
+                n++;
+            } while (File.Exists(codeBehindPath));
+            return codeBehindPath;
+        }
+
+        /// <summary>
+        /// Determine all explicit and implicit using statements in the code
+        /// document using the intermediate node.
+        /// </summary>
+        /// <param name="razorCodeDocument">The code document to analyze.</param>
+        /// <returns>An enumerable of the qualified namespaces.</returns>
+        private static IEnumerable<string> FindUsings(RazorCodeDocument razorCodeDocument)
+        {
+            return razorCodeDocument
+                .GetDocumentIntermediateNode()
+                .FindDescendantNodes<UsingDirectiveIntermediateNode>()
+                .Select(n => n.Content);
+        }
+
+        /// <summary>
+        /// Generate a complete C# compilation unit containing a partial class
+        /// with the given name, body contents, and the namespace and all
+        /// usings from the existing code document.
+        /// </summary>
+        /// <param name="className">Name of the resultant partial class.</param>
+        /// <param name="contents">Class body contents.</param>
+        /// <param name="razorCodeDocument">Existing code document we're extracting from.</param>
+        /// <returns></returns>
+        private static string GenerateCodeBehindClass(string className, string contents, RazorCodeDocument razorCodeDocument)
+        {
+            var namespaceNode = (NamespaceDeclarationIntermediateNode)razorCodeDocument
+                .GetDocumentIntermediateNode()
+                .FindDescendantNodes<IntermediateNode>()
+                .FirstOrDefault(n => n is NamespaceDeclarationIntermediateNode);
+            var namespaceName = namespaceNode?.Content ?? "Unknown";
```

Comment:
Hmm, I'd actually throw an error if you can't resolve the namespace. That should be a catastrophic error